File: app/controllers/admin/SettingsController.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.admin;

import be.objectify.deadbolt.java.actions.Group;
import be.objectify.deadbolt.java.actions.Restrict;
import be.objectify.deadbolt.java.actions.SubjectNotPresent;
import be.objectify.deadbolt.java.actions.SubjectPresent;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import controllers.base.ActionMethod;
import controllers.base.BaseController;
import io.ebean.DB;
import io.ebean.Update;
import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URI;
import java.net.URL;
import java.util.Optional;
import java.util.concurrent.CompletionStage;
import javax.inject.Inject;
import miscellaneous.config.ConfigReader;
import models.admin.GeneralSettings;
import models.enrolment.ExamEnrolment;
import models.user.Language;
import models.user.User;
import play.Environment;
import play.data.DynamicForm;
import play.libs.Json;
import play.libs.ws.WSClient;
import play.libs.ws.WSRequest;
import play.mvc.Http;
import play.mvc.Result;
import scala.jdk.javaapi.CollectionConverters;

public class SettingsController extends BaseController {

    private final Environment environment;
    private final ConfigReader configReader;
    private final WSClient wsClient;

    @Inject
    public SettingsController(Environment environment, ConfigReader configReader, WSClient wsClient) {
        this.environment = environment;
        this.configReader = configReader;
        this.wsClient = wsClient;
    }

    public static GeneralSettings get(String name) {
        return DB.find(GeneralSettings.class).where().eq("name", name).findOneOrEmpty().orElse(new GeneralSettings());
    }

    public static GeneralSettings getOrCreateSettings(String name, String value, String defaultValue) {
        GeneralSettings gs = DB.find(GeneralSettings.class).where().eq("name", name).findOne();
        if (gs == null) {
            gs = new GeneralSettings();
            gs.setName(name);
            gs.save();
        }
        if (value != null) {
            gs.setValue(value);
            gs.update();
        } else if (gs.getValue() == null && defaultValue != null) {
            gs.setValue(defaultValue);
            gs.update();
        }
        return gs;
    }

    @SubjectPresent
    public Result getUserAgreement() {
        GeneralSettings gs = getOrCreateSettings("eula", null, null);
        return ok(Json.toJson(gs));
    }

    @Restrict({ @Group("ADMIN"), @Group("STUDENT") })
    public Result getDeadline() {
        GeneralSettings gs = getOrCreateSettings("review_deadline", null, "14");
        return ok(Json.toJson(gs));
    }

    @Restrict({ @Group("ADMIN"), @Group("STUDENT") })
    public Result getReservationWindowSize() {
        GeneralSettings gs = getOrCreateSettings("reservation_window_size", null, "30");
        return ok(Json.toJson(gs));
    }

    @SubjectPresent
    public CompletionStage<Result> getMaturityInstructions(String lang, Optional<String> hash) throws IOException {
        Language language = DB.find(Language.class, lang);
        if (language == null) {
            return wrapAsPromise(badRequest("Language not supported"));
        }
        if (hash.isPresent()) {
            ExamEnrolment enrolment = DB.find(ExamEnrolment.class)
                .where()
                .eq("externalExam.hash", hash.get())
                .findOne();
            if (enrolment == null) {
                return wrapAsPromise(badRequest("Enrolment not found"));
            }
            URL url = parseExternalUrl(enrolment.getReservation().getExternalRef());
            WSRequest request = wsClient.url(url.toString()).addQueryParameter("lang", language.getCode());
            return request
                .get()
                .thenApplyAsync(response -> {
                    JsonNode root = response.asJson();
                    if (response.getStatus() != Http.Status.OK) {
                        return internalServerError(root.get("message").asText("Connection refused"));
                    }
                    return ok(root);
                });
        } else {
            String key = String.format("maturity_instructions_%s", lang);
            return wrapAsPromise(ok(Json.toJson(get(key))));
        }
    }

    @SubjectNotPresent
    public Result provideMaturityInstructions(String ref, String lang) {
        Language language = DB.find(Language.class, lang);
        if (language == null) {
            badRequest("Language not supported");
        }
        String key = String.format("maturity_instructions_%s", lang);
        return ok(Json.toJson(get(key)));
    }

    @Restrict({ @Group("ADMIN") })
    public Result updateUserAgreement(Http.Request request) {
        JsonNode body = request.body().asJson();
        String eula = body.get("value").asText();
        GeneralSettings gs = getOrCreateSettings("eula", eula, null);
        if (!body.get("minorUpdate").asBoolean()) {
            // Since the EULA has changed, force users to accept it again.
            String updStatement = "update app_user set user_agreement_accepted = :hasNot";
            Update<User> update = DB.createUpdate(User.class, updStatement);
            update.set("hasNot", false);
            update.execute();
        }
        return ok(Json.toJson(gs));
    }

    @Restrict({ @Group("ADMIN") })
    public Result setDeadline(Http.Request request) {
        DynamicForm df = formFactory.form().bindFromRequest(request);
        String deadline = df.get("value");
        GeneralSettings gs = getOrCreateSettings("review_deadline", deadline, null);
        return ok(Json.toJson(gs));
    }

    @Restrict({ @Group("ADMIN") })
    public Result setReservationWindowSize(Http.Request request) {
        DynamicForm df = formFactory.form().bindFromRequest(request);
        String deadline = df.get("value");
        GeneralSettings gs = getOrCreateSettings("reservation_window_size", deadline, null);
        return ok(Json.toJson(gs));
    }

    @SubjectPresent
    public Result getHostname() {
        ObjectNode node = Json.newObject();
        node.put("hostname", configReader.getHostName());
        return ok(Json.toJson(node));
    }

    @SubjectPresent
    public Result getMaxFilesize() {
        ObjectNode node = Json.newObject();
        node.put("filesize", configReader.getMaxFileSize());
        return ok(Json.toJson(node));
    }

    @Restrict({ @Group("ADMIN"), @Group("SUPPORT"), @Group("TEACHER") })
    public Result getExamDurations() {
        ObjectNode node = Json.newObject();
        ArrayNode durations = node.putArray("examDurations");
        configReader.getExamDurationsJava().forEach(durations::add);
        return ok(Json.toJson(node));
    }

    @Restrict({ @Group("ADMIN"), @Group("SUPPORT"), @Group("TEACHER") })
    public Result getExamMaxDuration() {
        ObjectNode node = Json.newObject();
        node.put("maxDuration", configReader.getExamMaxDuration());
        return ok(Json.toJson(node));
    }

    @Restrict({ @Group("ADMIN"), @Group("SUPPORT"), @Group("TEACHER") })
    public Result getExamMinDuration() {
        ObjectNode node = Json.newObject();
        node.put("minDuration", configReader.getExamMinDuration());
        return ok(Json.toJson(node));
    }

    @Restrict({ @Group("ADMIN"), @Group("SUPPORT"), @Group("TEACHER") })
    public Result isExamGradeScaleOverridable() {
        ObjectNode node = Json.newObject();
        node.put("overridable", configReader.isCourseGradeScaleOverridable());
        return ok(Json.toJson(node));
    }

    @Restrict({ @Group("ADMIN"), @Group("TEACHER"), @Group("STUDENT") })
    public Result isEnrolmentPermissionCheckActive() {
        ObjectNode node = Json.newObject();
        node.put("active", configReader.isEnrolmentPermissionCheckActive());
        return ok(Json.toJson(node));
    }

    @Restrict({ @Group("ADMIN") })
    public Result getAppVersion() {
        ObjectNode node = Json.newObject();
        node.put("appVersion", configReader.getAppVersion());
        return ok(Json.toJson(node));
    }

    @ActionMethod
    public Result isProd() {
        ObjectNode node = Json.newObject();
        node.put("isProd", environment.isProd());
        return ok(Json.toJson(node));
    }

    @ActionMethod
    public Result isExamVisitSupported() {
        ObjectNode node = Json.newObject();
        node.put("isExamVisitSupported", configReader.isVisitingExaminationSupported());
        return ok(Json.toJson(node));
    }

    @ActionMethod
    public Result isExamCollaborationSupported() {
        ObjectNode node = Json.newObject();
        node.put("isExamCollaborationSupported", configReader.isCollaborationExaminationSupported());
        return ok(Json.toJson(node));
    }

    @ActionMethod
    public Result isAnonymousReviewEnabled() {
        ObjectNode node = Json.newObject();
        node.put("anonymousReviewEnabled", configReader.isAnonymousReviewEnabled());
        return ok(Json.toJson(node));
    }

    @ActionMethod
    public Result getByodSupport() {
        ObjectNode node = Json.newObject();
        node
            .put("sebExaminationSupported", configReader.isSebExaminationSupported())
            .put("homeExaminationSupported", configReader.isHomeExaminationSupported());
        return ok(Json.toJson(node));
    }

    @Restrict({ @Group("STUDENT") })
    public Result getExaminationQuitLink() {
        ObjectNode node = Json.newObject();
        node.put("quitLink", configReader.getQuitExaminationLink());
        return ok(Json.toJson(node));
    }

    @Restrict({ @Group("ADMIN") })
    public Result getConfig() {
        ObjectNode node = Json.newObject();

        node.put("hasCourseSearchIntegration", configReader.isCourseSearchActive());
        node.put("anonymousReviewEnabled", configReader.isAnonymousReviewEnabled());
        ObjectNode courseIntegrationUrls = Json.newObject();
        CollectionConverters.asJava(configReader.getCourseIntegrationUrls()).forEach(courseIntegrationUrls::put);
        node.set("courseSearchIntegrationUrls", courseIntegrationUrls);

        ArrayNode durations = Json.newArray();
        configReader.getExamDurationsJava().forEach(durations::add);
        node.set("examDurations", durations);

        ObjectNode roles = Json.newObject();
        configReader
            .getRoleMappingJava()
            .forEach((k, v) -> {
                ArrayNode role = Json.newArray();
                v.forEach(role::add);
                roles.set(k.getName(), role);
            });
        node.set("roles", roles);

        GeneralSettings eula = getOrCreateSettings("eula", null, null);
        node.put("eula", eula.getValue());
        GeneralSettings reservationWindowSize = getOrCreateSettings("reservation_window_size", null, "30");
        node.put("reservationWindowSize", Integer.parseInt(reservationWindowSize.getValue()));
        GeneralSettings reviewDeadline = getOrCreateSettings("review_deadline", null, "14");
        node.put("reviewDeadline", Integer.parseInt(reviewDeadline.getValue()));

        node.put("isExamVisitSupported", configReader.isVisitingExaminationSupported());
        node.put("isExamCollaborationSupported", configReader.isCollaborationExaminationSupported());
        node.put("hasEnrolmentCheckIntegration", configReader.isEnrolmentPermissionCheckActive());
        node.put("isGradeScaleOverridable", configReader.isCourseGradeScaleOverridable());
        node.put("supportsMaturity", configReader.isMaturitySupported());
        node.put("supportsPrintouts", configReader.isPrintoutSupported());
        node.put("maxFileSize", configReader.getMaxFileSize());
        node.put("expirationPeriod", configReader.getExamExpirationPeriod());
        node.put("defaultTimeZone", configReader.getDefaultTimeZone().getID());
        node.put("sebQuitLink", configReader.getQuitExaminationLink());
        node.put("isSebExaminationSupported", configReader.isSebExaminationSupported());
        node.put("isHomeExaminationSupported", configReader.isHomeExaminationSupported());

        return ok(Json.toJson(node));
    }

    @SubjectPresent
    public Result getCourseCodePrefix() {
        ObjectNode node = Json.newObject();
        node.put("prefix", configReader.getCourseCodePrefix());
        return ok(Json.toJson(node));
    }

    @Restrict({ @Group("ADMIN"), @Group("SUPPORT"), @Group("TEACHER") })
    public Result getByodMaxParticipants() {
        ObjectNode node = Json.newObject();
        node.put("max", configReader.getMaxByodExaminationParticipantCount());
        return ok(Json.toJson(node));
    }

    private URL parseExternalUrl(String reservationRef) throws MalformedURLException {
        return URI.create(
            configReader.getIopHost() + String.format("/api/enrolments/%s/instructions", reservationRef)
        ).toURL();
    }
}

File: app/controllers/admin/SettingsController.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.admin

import io.ebean.DB
import miscellaneous.config.ConfigReader
import miscellaneous.scala.{DbApiHelper, JavaApiHelper}
import models.admin.GeneralSettings
import models.enrolment.ExamEnrolment
import models.user.{Language, Role, User}
import play.api.libs.json.*
import play.api.libs.ws.WSClient
import play.api.mvc.*
import security.scala.Auth.{AuthenticatedAction, authorized}
import security.scala.AuthExecutionContext
import system.AuditedAction

import java.net.URI
import javax.inject.Inject
import scala.concurrent.Future
import scala.util.Try

class SettingsController @Inject() (
    val controllerComponents: ControllerComponents,
    val authenticated: AuthenticatedAction,
    val audited: AuditedAction,
    val configReader: ConfigReader,
    val wsClient: WSClient,
    implicit val ec: AuthExecutionContext
) extends BaseController
    with JavaApiHelper
    with DbApiHelper:

  // Helper methods
  private def get(name: String): GeneralSettings =
    DB.find(classOf[GeneralSettings])
      .where()
      .eq("name", name)
      .find
      .getOrElse(new GeneralSettings())

  // Endpoints
  def getUserAgreement: Action[AnyContent] = authenticated { _ =>
    Ok(configReader.getOrCreateSettings("eula", None, None).asJson)
  }

  def getDeadline: Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN, Role.Name.STUDENT))) { _ =>
      Ok(configReader.getOrCreateSettings("review_deadline", None, Some("14")).asJson)
    }

  def getReservationWindowSize: Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN, Role.Name.STUDENT))) { _ =>
      Ok(configReader.getOrCreateSettings("reservation_window_size", None, Some("30")).asJson)
    }

  def getMaturityInstructions(lang: String, hash: Option[String]): Action[AnyContent] = authenticated.async { _ =>
    Option(DB.find(classOf[Language], lang)) match
      case None => Future.successful(BadRequest("Language not supported"))
      case Some(language) =>
        hash match
          case Some(h) =>
            DB.find(classOf[ExamEnrolment])
              .where()
              .eq("externalExam.hash", h)
              .find match
              case None => Future.successful(BadRequest("Enrolment not found"))
              case Some(enrolment) =>
                Try(parseExternalUrl(enrolment.getReservation.getExternalRef)).toOption match
                  case None => Future.successful(BadRequest("Invalid external URL"))
                  case Some(url) =>
                    wsClient
                      .url(url.toString)
                      .addQueryStringParameters("lang" -> language.getCode)
                      .get()
                      .map { response =>
                        if response.status == OK then Ok(response.json)
                        else
                          InternalServerError(
                            (response.json \ "message").asOpt[String].getOrElse("Connection refused")
                          )
                      }
          case None =>
            val key = s"maturity_instructions_$lang"
            Future.successful(Ok(get(key).asJson))
  }

  def provideMaturityInstructions(ref: String, lang: String): Action[AnyContent] = Action { _ =>
    Option(DB.find(classOf[Language], lang)) match
      case None => BadRequest("Language not supported")
      case Some(_) =>
        val key = s"maturity_instructions_$lang"
        Ok(get(key).asJson)
  }

  def updateUserAgreement(): Action[JsValue] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))).andThen(audited)(parse.json) { request =>
      val eula = (request.body \ "value").as[String]
      val gs   = configReader.getOrCreateSettings("eula", Some(eula), None)

      if !(request.body \ "minorUpdate").as[Boolean] then
        // Force users to accept EULA again
        val update = DB.createUpdate(classOf[User], "update app_user set user_agreement_accepted = :hasNot")
        update.set("hasNot", false)
        update.execute()

      Ok(gs.asJson)
    }

  def setDeadline(): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))).andThen(audited) { request =>
      request.body.asFormUrlEncoded.flatMap(_.get("value").flatMap(_.headOption)) match
        case Some(deadline) =>
          Ok(configReader.getOrCreateSettings("review_deadline", Some(deadline), None).asJson)
        case None => BadRequest("Missing value")
    }

  def setReservationWindowSize(): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))).andThen(audited) { request =>
      request.body.asFormUrlEncoded.flatMap(_.get("value").flatMap(_.headOption)) match
        case Some(size) =>
          Ok(configReader.getOrCreateSettings("reservation_window_size", Some(size), None).asJson)
        case None => BadRequest("Missing value")
    }

  def getHostname: Action[AnyContent] = authenticated { _ =>
    Ok(Json.obj("hostname" -> configReader.getHostName))
  }

  def getMaxFilesize: Action[AnyContent] = authenticated { _ =>
    Ok(Json.obj("filesize" -> configReader.getMaxFileSize))
  }

  def getExamDurations: Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN, Role.Name.SUPPORT, Role.Name.TEACHER))) { _ =>
      Ok(Json.obj("examDurations" -> configReader.getExamDurations))
    }

  def getExamMaxDuration: Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN, Role.Name.SUPPORT, Role.Name.TEACHER))) { _ =>
      Ok(Json.obj("maxDuration" -> configReader.getExamMaxDuration))
    }

  def getExamMinDuration: Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN, Role.Name.SUPPORT, Role.Name.TEACHER))) { _ =>
      Ok(Json.obj("minDuration" -> configReader.getExamMinDuration))
    }

  def isExamGradeScaleOverridable: Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN, Role.Name.SUPPORT, Role.Name.TEACHER))) { _ =>
      Ok(Json.obj("overridable" -> configReader.isCourseGradeScaleOverridable))
    }

  def isEnrolmentPermissionCheckActive: Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN, Role.Name.TEACHER, Role.Name.STUDENT))) { _ =>
      Ok(Json.obj("active" -> configReader.isEnrolmentPermissionCheckActive))
    }

  def getAppVersion: Action[AnyContent] = authenticated.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
    Ok(Json.obj("appVersion" -> configReader.getAppVersion))
  }

  def isProd: Action[AnyContent] = Action { _ =>
    Ok(Json.obj("isProd" -> (configReader.getLoginType != "DEBUG")))
  }

  def isExamVisitSupported: Action[AnyContent] = Action { _ =>
    Ok(Json.obj("isExamVisitSupported" -> configReader.isVisitingExaminationSupported))
  }

  def isExamCollaborationSupported: Action[AnyContent] = Action { _ =>
    Ok(Json.obj("isExamCollaborationSupported" -> configReader.isCollaborationExaminationSupported))
  }

  def isAnonymousReviewEnabled: Action[AnyContent] = Action { _ =>
    Ok(Json.obj("anonymousReviewEnabled" -> configReader.isAnonymousReviewEnabled))
  }

  def getByodSupport: Action[AnyContent] = Action { _ =>
    Ok(
      Json.obj(
        "sebExaminationSupported"  -> configReader.isSebExaminationSupported,
        "homeExaminationSupported" -> configReader.isHomeExaminationSupported
      )
    )
  }

  def getExaminationQuitLink: Action[AnyContent] = authenticated.andThen(authorized(Seq(Role.Name.STUDENT))) { _ =>
    Ok(Json.obj("quitLink" -> configReader.getQuitExaminationLink))
  }

  def getConfig: Action[AnyContent] = authenticated.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
    val courseIntegrationUrls = JsObject(configReader.getCourseIntegrationUrls.toSeq.map { case (k, v) =>
      k -> JsString(v)
    })

    val roles = JsObject(
      configReader.getRoleMapping.map { case (k, v) => k.getName -> JsArray(v.map(JsString(_))) }.toSeq
    )

    val eula                  = configReader.getOrCreateSettings("eula", None, None)
    val reservationWindowSize = configReader.getOrCreateSettings("reservation_window_size", None, Some("30"))
    val reviewDeadline        = configReader.getOrCreateSettings("review_deadline", None, Some("14"))

    Ok(
      Json.obj(
        "hasCourseSearchIntegration"   -> configReader.isCourseSearchActive,
        "anonymousReviewEnabled"       -> configReader.isAnonymousReviewEnabled,
        "courseSearchIntegrationUrls"  -> courseIntegrationUrls,
        "examDurations"                -> configReader.getExamDurations,
        "roles"                        -> roles,
        "eula"                         -> eula.getValue,
        "reservationWindowSize"        -> reservationWindowSize.getValue.toInt,
        "reviewDeadline"               -> reviewDeadline.getValue.toInt,
        "isExamVisitSupported"         -> configReader.isVisitingExaminationSupported,
        "isExamCollaborationSupported" -> configReader.isCollaborationExaminationSupported,
        "hasEnrolmentCheckIntegration" -> configReader.isEnrolmentPermissionCheckActive,
        "isGradeScaleOverridable"      -> configReader.isCourseGradeScaleOverridable,
        "supportsMaturity"             -> configReader.isMaturitySupported,
        "supportsPrintouts"            -> configReader.isPrintoutSupported,
        "maxFileSize"                  -> configReader.getMaxFileSize,
        "expirationPeriod"             -> configReader.getExamExpirationPeriod,
        "defaultTimeZone"              -> configReader.getDefaultTimeZone.getID,
        "sebQuitLink"                  -> configReader.getQuitExaminationLink,
        "isSebExaminationSupported"    -> configReader.isSebExaminationSupported,
        "isHomeExaminationSupported"   -> configReader.isHomeExaminationSupported
      )
    )
  }

  def getCourseCodePrefix: Action[AnyContent] = authenticated { _ =>
    Ok(Json.obj("prefix" -> configReader.getCourseCodePrefix))
  }

  def getByodMaxParticipants: Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN, Role.Name.SUPPORT, Role.Name.TEACHER))) { _ =>
      Ok(Json.obj("max" -> configReader.getMaxByodExaminationParticipantCount))
    }

  private def parseExternalUrl(reservationRef: String): java.net.URL =
    URI
      .create(
        configReader.getIopHost + s"/api/enrolments/$reservationRef/instructions"
      )
      .toURL
